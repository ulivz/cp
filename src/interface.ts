/**
 * Transform file.
 */
export type TransformerType = (
  this: ICopyOptions,
  content: string,
  filename: string
) => string;
/**
 * Rename file.
 */
export type RenamerType = (originalName: string) => string;
/**
 * Set default content, it will be used when the source
 * file doesn't exist.
 */
export type OverrideFnType = (filename: string, content?: string) => string;

/**
 * A inteface to describe file action.
 */
export interface IFileObjectDescriptor {
  /**
   * Transform the file.
   */
  transform?: TransformerType;
  /**
   * Rename the source file.
   */
  rename?: string | RenamerType;
  /**
   * Whether to skip existence check when the source file
   * doesn't exist, Note that the check will be disabled
   * when the 'default' content is given.
   */
  skipIfNotExists?: boolean;
  /**
   * Override content.
   * Note that if the source files don't exists, the overriding
   * result will be used as the default content.
   */
  override?: string | OverrideFnType;
  /**
   * Whether to ignore this file.
   */
  ignored?: boolean;
}

/**
 * Combined descriptor.
 */
export type IFileDescriptor =
  | IFileObjectDescriptor
  | boolean
  | string
  | OverrideFnType;

/**
 * Array item for usage of descriptors.
 */
export type IFileDescriptorArrayItem = string | [string, IFileDescriptor?];

/**
 * Array usage of descriptors for users.
 */
export type IFileDescriptorArray = Array<IFileDescriptorArrayItem>;

/**
 * Map usage of descriptors for users.
 */
export interface IFileDescriptorMap {
  [pattern: string]: IFileDescriptor;
}

/**
 * Descriptor interface for users.
 */
export type IFileDescriptors = IFileDescriptorArray | IFileDescriptorMap;

/**
 * Normalized descriptor.
 */
export type INomalizedDescriptor = [
  string /* pattern */,
  IFileObjectDescriptor
];

/**
 * Normalized descriptors.
 */
export type INomalizedDescriptors = Array<INomalizedDescriptor>;

/**
 * User options.
 */
export interface ICopyOptions {
  /**
   * Source directory.
   */
  src: string;
  /**
   * Output directory.
   */
  dist: string;
  /**
   * Files handler.
   */
  files?: IFileDescriptors;
  /**
   * Debug flag.
   */
  debug?: boolean;
  /**
   * Whether to write to disk.
   */
  write?: boolean;
  /**
   * Global skipIfNotExists
   */
  skipIfNotExists?: boolean;
  /**
   * Disable the "override" option under file descriptor.
   * the overriding content will be fallback to default content.
   */
  disableOverride?: boolean;
  /**
   * Enable migrate mode, it will clean source files that have been renamed.
   *
   * NOTE: It only takes effect when `src` is same to `dist`.
   */
  enableMigrateMode?: boolean;
}

export interface IFilesMap {
  [pattern: string]: string;
}
