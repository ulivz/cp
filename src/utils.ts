/**
 * @license
 * Copyright (c) ULIVZ. All Rights Reserved.
 */

import { File } from "majo";

/**
 * Get the raw type string of a value, e.g., [object Object].
 */
const _toString = Object.prototype.toString;

/**
 * Strict object type check. Only returns true for plain JavaScript objects.
 */
export function isPlainObject(obj: unknown): boolean {
  return _toString.call(obj) === "[object Object]";
}

/**
 * Create a majo file.
 */
export function createMajoFileSync(filepath: string, contents: string): File {
  return {
    contents: Buffer.from(contents),
    stats: null,
    path: filepath,
  } as File;
}
