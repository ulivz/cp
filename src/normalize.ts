import { isPlainObject } from "./utils";
import {
  IFileDescriptor,
  IFileDescriptorArray,
  IFileDescriptorMap,
  IFileDescriptors,
  IFileObjectDescriptor,
  INomalizedDescriptors,
} from "./interface";

export function normalizeFileDescriptor(
  descriptor: IFileDescriptor
): IFileObjectDescriptor {
  if (isPlainObject(descriptor)) {
    return descriptor as IFileObjectDescriptor;
  }
  if (typeof descriptor === "string" || typeof descriptor === "function") {
    return { override: descriptor };
  }
  if (descriptor === true) {
    return {};
  }
  if (descriptor === false) {
    return { ignored: true };
  }
  throw new Error(`Invalid file descriptor: ${descriptor}`);
}

/**
 * Normalize @IFileDescriptors to @INomalizedDescriptors
 */
export function normalizeFileDescriptors(
  descriptors: IFileDescriptors
): INomalizedDescriptors | null {
  if (Array.isArray(descriptors)) {
    return (descriptors as IFileDescriptorArray).map((descriptor) => {
      descriptor = Array.isArray(descriptor) ? descriptor : [descriptor];

      const [pattern, fileDescriptor = {}] = descriptor;

      if (typeof pattern !== "string") {
        throw new Error(`Invalid file pattern: ${pattern}`);
      }

      const objectDescriptor = normalizeFileDescriptor(fileDescriptor);

      return [pattern, objectDescriptor];
    });
  }

  if (isPlainObject(descriptors)) {
    return Object.keys(descriptors as IFileDescriptorMap).map(
      (pattern: string) => {
        const descriptor = descriptors[pattern];
        const objectDescriptor = normalizeFileDescriptor(descriptor);

        return [pattern, objectDescriptor];
      }
    );
  }

  return null;
}
