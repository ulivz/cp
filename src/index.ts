/**
 * @license
 * Copyright (c) ULIVZ. All Rights Reserved.
 */

import * as path from "path";
import * as os from "os";
import * as fs from "fs-extra";
import tildify from "tildify";
import matcher from "micromatch";
import { majo, Majo } from "majo";
import { createMajoFileSync } from "./utils";
import { normalizeFileDescriptors } from "./normalize";

export * from "./interface";
import {
  INormalizedDescriptor,
  INormalizedDescriptors,
  ICopyOptions,
  IFilesMap,
} from "./interface";

const debug = require("debug")("cp");

declare module "majo" {
  interface Majo {
    fileMap: () => IFilesMap;
    srcFileMap: () => IFilesMap;
    _srcFileMap: IFilesMap;
    destBaseDir: string;
    existsSync(filename: string): boolean;
  }
}

Majo.prototype.fileMap = function () {
  return getFilesMap(this);
};

Majo.prototype.srcFileMap = function () {
  return this._srcFileMap;
};

export { majo, Majo };

export const TEMP = "TEMP";

export { Majo as ICopyStream };

const MEDIA_FILE_REG = /^.*\.(jpg|png|webp|mov|mp4|gif|doc|pdf)$/i;

export async function cp(opts: ICopyOptions): Promise<Majo> {
  const {
    src: sourceDir,
    files: fileDescriptors = ["**"],
    debug: displayLog = false,
    write: writeToDisk = true,
    skipIfNotExists: globalSkipIfNotExists = false,
    disableOverride = false,
    enableMigrateMode = false,
  } = opts;

  let { dist: targetDir } = opts;

  if (targetDir === TEMP) {
    targetDir = path.join(
      os.tmpdir(),
      new Date().getTime().toString(36) + Math.random().toString(36)
    );
    fs.ensureDirSync(targetDir);
  }

  if (displayLog) {
    require("debug").enable("cp");
  }

  /**
   * Normalize snippets
   */
  const nFileDescriptors: INormalizedDescriptors =
    normalizeFileDescriptors(fileDescriptors);
  if (nFileDescriptors === null) {
    throw new Error(`Invalid value for "files" option: ${fileDescriptors}`);
  }

  const normalize =
    sourceDir === targetDir ? (filepath: string): string => filepath : tildify;

  /**
   * A majo middleware to cache source files.
   */
  function cacheSourcefilesMiddleware(stream: Majo): void {
    stream._srcFileMap = getFilesMap(stream);
  }

  /**
   * A majo middleware to handle core functions for cp.
   */
  async function coreMiddleware(stream: Majo): Promise<void> {
    await Promise.all(nFileDescriptors.map(async (descriptor: INormalizedDescriptor) => {
      const [
        pattern,
        {
          transform,
          rename,
          skipIfNotExists = false,
          ignored,
          override: overrideGetter,
        },
      ] = descriptor;

      debug("enter pattern", pattern);

      if (ignored) {
        return;
      }

      const { fileList } = stream;
      const targetFiles = matcher(fileList, pattern);

      /**
       * Handle that source files doesn't exists.
       */
      if (targetFiles.length === 0) {
        const abFilepath = path.join(sourceDir, pattern);

        if (overrideGetter) {
          let defaultContent = overrideGetter;
          if (typeof overrideGetter === "function") {
            defaultContent = overrideGetter(pattern);
          }

          debug("Created", normalize(pattern));
          stream.createFile(
            pattern,
            createMajoFileSync(abFilepath, defaultContent as string)
          );
        } else if (skipIfNotExists || globalSkipIfNotExists) {
          return;
        } else {
          throw new Error(`"${pattern}" not exist！`);
        }

        return;
      }

      /**
       * Handle that source files exists
       */
      await Promise.all(targetFiles.map(async (filename: string) => {
        if (MEDIA_FILE_REG.test(filename)) {
          return;
        }
        let fileContent = stream.fileContents(filename);

        if (rename) {
          const outputFilename =
            typeof rename === "function"
              ? rename(filename)
              : typeof rename === "string"
              ? rename
              : filename;

          /**
           * Only apply renaming when the target filename isn't same to filename.
           * Or the new file will be deleted after it was created.
           */
          if (filename !== outputFilename) {
            stream.rename(filename, outputFilename);
            filename = outputFilename;
          }
        }

        if (typeof transform === "function") {
          fileContent = await transform.call(opts, fileContent, filename);
          debug("Transformed", normalize(filename));
          stream.writeContents(filename, fileContent as string);
        }

        if (overrideGetter && !disableOverride) {
          let overrideContent = overrideGetter;
          if (typeof overrideGetter === "function") {
            overrideContent = overrideGetter(pattern, fileContent);
          }
          stream.writeContents(filename, overrideContent as string);
        }

        debug("copied", normalize(filename));
      }));
    }));
  }

  const globalPatterns = nFileDescriptors.map((descriptor) => {
    const [pattern, { ignored }] = descriptor;
    return ignored ? `!${pattern}` : pattern;
  });

  debug("globalPatterns", globalPatterns);

  const stream = majo();
  await stream
    .source(globalPatterns, { baseDir: sourceDir })
    .use(cacheSourcefilesMiddleware)
    .use(coreMiddleware);

  /**
   * Handle `write`
   */
  if (writeToDisk) {
    await stream.dest("", { baseDir: targetDir });
  } else {
    await stream.process();
  }

  /**
   * Handle `enableMigrateMode`
   */
  if (enableMigrateMode && sourceDir === targetDir) {
    const srcFileMap = stream.srcFileMap();
    const distFileMap = stream.fileMap();
    const srcFilePaths = Object.keys(srcFileMap);
    await Promise.all(
      srcFilePaths.map(async (srcFilePath) => {
        const absFilePath = path.join(sourceDir, srcFilePath);
        if (!distFileMap[srcFilePath] && fs.existsSync(absFilePath)) {
          await fs.remove(absFilePath);
        }
      })
    );
  }

  stream.destBaseDir = targetDir;
  stream.existsSync = (filename: string) => {
    return !!stream.file(filename);
  };

  return stream;
}

function getFilesMap(stream: Majo) {
  return stream.fileList.reduce<IFilesMap>((memo, filename) => {
    memo[filename] = stream.fileContents(filename);
    return memo;
  }, {});
}
