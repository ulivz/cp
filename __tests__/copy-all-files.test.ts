import { cp } from '../src';
import { src, dist } from './util';

describe('cp all files', () => {
  it('default usage should be equivalent to array and object usage.', async () => {
    const stream1 = await cp({
      src,
      dist,
      write: false,
      debug: true,
    });

    expect(stream1.fileMap()).toMatchSnapshot();

    const stream2 = await cp({
      src,
      dist,
      write: false,
      debug: true,
      files: ['**'],
    });

    expect(stream2.fileMap()).toEqual(stream1.fileMap());

    const stream3 = await cp({
      src,
      dist,
      write: false,
      debug: true,
      files: {
        '**': true,
      },
    });

    expect(stream3.fileMap()).toEqual(stream1.fileMap());
  });

  it('source dir is same to out dir', async () => {
    const stream = await cp({
      src,
      dist: src,
      debug: true,
      write: false,
    });

    expect(stream.fileList.length).toBe(3);
  });
});
