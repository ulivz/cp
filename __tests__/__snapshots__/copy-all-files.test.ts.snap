// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copy all files default usage should be equalivant to array and object usage. 1`] = `
Object {
  "package.json": "{
  \\"name\\": \\"copy\\"
}
",
  "src/bin.ts": "export function bin() {

}
",
  "src/index.ts": "import { bin } from './bin';

export default bin;
",
}
`;

exports[`cp all files default usage should be equivalent to array and object usage. 1`] = `
Object {
  "package.json": "{
  \\"name\\": \\"copy\\"
}
",
  "src/bin.ts": "export function bin() {

}
",
  "src/index.ts": "import { bin } from './bin';

export default bin;
",
}
`;
